File: ios-native/Dimo/Features/SignIn/SignInScreen.swift
import SwiftUI

struct SignInScreen: View {
  @Environment(AppEnvironment.self) private var environment
  @State private var signingInWith: AuthProviderKind?
  @State private var errorMessage: String?

  var body: some View {
    VStack(spacing: 28) {
      Spacer()
      ZStack {
        RoundedRectangle(cornerRadius: 28, style: .continuous)
          .fill(Theme.green)
          .frame(width: 88, height: 88)
        Text("D")
          .font(DimoFont.display(44, weight: .bold))
          .foregroundStyle(Theme.onGreen)
      }
      VStack(spacing: 8) {
        Text("Welcome to Dimo")
          .font(DimoFont.display(28, weight: .bold))
          .foregroundStyle(Theme.ink)
        Text("Track spending with a calm, local-first ledger.")
          .font(DimoFont.body(15))
          .foregroundStyle(Theme.muted)
          .multilineTextAlignment(.center)
          .padding(.horizontal, 36)
      }
      if let errorMessage {
        Text(errorMessage)
          .font(DimoFont.body(13))
          .foregroundStyle(Theme.danger)
          .multilineTextAlignment(.center)
          .padding(.horizontal, 24)
      }
      Spacer()
      VStack(spacing: 12) {
        // Apple first: Sign in with Apple must be shown no less prominently
        // than other providers.
        ProviderButton(
          title: "Sign in with Apple",
          busyTitle: "Signing in…",
          icon: .symbol("applelogo"),
          background: Theme.ink,
          foreground: Theme.canvas,
          isBusy: signingInWith == .apple,
          enabled: signingInWith == nil
        ) {
          Task { await signIn(with: .apple) }
        }
        ProviderButton(
          title: "Sign in with Google",
          busyTitle: "Signing in…",
          icon: .asset("GoogleG"),
          background: Theme.green,
          foreground: Theme.onGreen,
          isBusy: signingInWith == .google,
          enabled: signingInWith == nil
        ) {
          Task { await signIn(with: .google) }
        }
      }
      .padding(.horizontal, 24)
      .padding(.bottom, 40)
    }
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(Theme.canvas.ignoresSafeArea())
  }

  private func signIn(with provider: AuthProviderKind) async {
    signingInWith = provider
    errorMessage = nil
    defer { signingInWith = nil }
    do {
      try await environment.session.signIn(with: provider)
    } catch AuthError.cancelled {
      // Dismissing the web sheet is not an error worth shouting about.
      errorMessage = nil
    } catch {
      errorMessage = error.localizedDescription
    }
  }
}

/// Provider mark shown ahead of the title. Asset marks render as templates so
/// they pick up the button's foreground colour.
private enum ProviderIcon {
  case symbol(String)
  case asset(String)
}

private struct ProviderButton: View {
  var title: String
  var busyTitle: String
  var icon: ProviderIcon?
  var background: Color
  var foreground: Color
  var isBusy: Bool
  var enabled: Bool
  var action: () -> Void

  var body: some View {
    Button(action: action) {
      HStack(spacing: 10) {
        if isBusy {
          ProgressView()
            .tint(foreground)
        } else {
          switch icon {
          case .symbol(let name):
            Image(systemName: name)
              .font(.system(size: 17, weight: .medium))
          case .asset(let name):
            Image(name)
              .renderingMode(.template)
              .resizable()
              .scaledToFit()
              .frame(width: 18, height: 18)
          case nil:
            EmptyView()
          }
        }
        Text(isBusy ? busyTitle : title)
          .font(DimoFont.body(16, weight: .semibold))
      }
      .frame(maxWidth: .infinity)
      .padding(.vertical, 16)
      .background(background)
      .foregroundStyle(foreground)
      .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
    }
    .buttonStyle(.plain)
    .disabled(!enabled)
    .opacity(enabled || isBusy ? 1 : 0.55)
  }
}

File: android-native/app/src/main/java/app/dimo/android/features/signin/SignInScreen.kt
package app.dimo.android.features.signin

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.dimo.android.app.LocalAppEnvironment
import app.dimo.android.design.ActionButton
import app.dimo.android.design.ActionButtonVariant
import app.dimo.android.design.DimoColors
import app.dimo.android.design.DimoFont
import kotlinx.coroutines.launch

/** Port of `ios-native/Dimo/Features/SignIn/SignInScreen.swift`. */
@Composable
fun SignInScreen(modifier: Modifier = Modifier) {
  val environment = LocalAppEnvironment.current
  val scope = rememberCoroutineScope()
  var isSigningIn by remember { mutableStateOf(false) }
  var errorMessage by remember { mutableStateOf<String?>(null) }

  Column(
    modifier = modifier
      .fillMaxSize()
      .background(DimoColors.canvas)
      .padding(horizontal = 24.dp),
    horizontalAlignment = Alignment.CenterHorizontally,
  ) {
    Spacer(modifier = Modifier.weight(1f))
    Box(
      modifier = Modifier
        .size(88.dp)
        .clip(RoundedCornerShape(28.dp))
        .background(DimoColors.green),
      contentAlignment = Alignment.Center,
    ) {
      Text(
        text = "D",
        style = DimoFont.display(44f, FontWeight.Bold),
        color = DimoColors.onGreen,
      )
    }
    Spacer(modifier = Modifier.height(28.dp))
    Column(
      horizontalAlignment = Alignment.CenterHorizontally,
      verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
      Text(
        text = "Welcome to Dimo",
        style = DimoFont.display(28f, FontWeight.Bold),
        color = DimoColors.ink,
      )
      Text(
        text = "Track spending with a calm, local-first ledger.",
        style = DimoFont.body(15f),
        color = DimoColors.muted,
        textAlign = TextAlign.Center,
        modifier = Modifier.padding(horizontal = 12.dp),
      )
    }
    if (errorMessage != null) {
      Spacer(modifier = Modifier.height(16.dp))
      Text(
        text = errorMessage.orEmpty(),
        style = DimoFont.body(13f),
        color = DimoColors.danger,
        textAlign = TextAlign.Center,
      )
    }
    Spacer(modifier = Modifier.weight(1f))
    ActionButton(
      title = if (isSigningIn) "Signing in…" else "Continue with Google",
      onClick = {
        if (isSigningIn) return@ActionButton
        scope.launch {
          isSigningIn = true
          errorMessage = null
          try {
            environment.session.signInWithGoogle()
          } catch (error: Exception) {
            if (error.message != "Sign-in cancelled") {
              errorMessage = error.message ?: "Sign-in failed"
            }
          } finally {
            isSigningIn = false
          }
        }
      },
      variant = ActionButtonVariant.Accent,
      enabled = !isSigningIn,
      modifier = Modifier.fillMaxWidth(),
    )
    if (isSigningIn) {
      Spacer(modifier = Modifier.height(12.dp))
      CircularProgressIndicator(
        color = DimoColors.green,
        strokeWidth = 2.dp,
        modifier = Modifier.size(22.dp),
      )
    }
    Spacer(modifier = Modifier.height(40.dp))
  }
}
